Sessions: Gutenberg-ify Session Info & Speakers metaboxes

ryelle opened the pull request:
This is an attempt at switching over to using a gutenberg pattern for the session meta info (date, time, etc). It sets up a gutenberg plugin which slots into the Document Settings sidebar, and creates panels for Session Info & Speakers.

The idea is that we could phase out the metabox & meta saving functionality, and instead expose the meta via API + `register_post_meta` and use the UI patterns of Gutenberg.

### Screenshots

![Screen Shot 2020-01-14 at 4 56 17 PM](https://user-images.githubusercontent.com/541093/72386997-820a8580-36f0-11ea-9239-d7b7aca408b3.png)

### How to test the changes in this Pull Request:

1. Edit a session (either new, or existing)
2. If existing, the session info should correctly populate the new panel
3. If new, it should be empty
4. Edit the session info, save
5. It should save correctly

ryelle committed:
Post Types: Create a yarn workspace for wc-post-types code

ryelle committed:
Sessions: Use a plugin panel in the Document sidebar for Session Info

This removes the Session Info metabox, in favor of a native gutenberg sidebar panel. Now we can use core form components, including the datepicker for the session date.

ryelle committed:
Sessions: Remove saving logic from the save_post hook, since the metabox does not exist anymore

This also adds the wp.tv validation to the video meta callback, so API data is also correctly validated.

ryelle commented on public_html/wp-content/plugins/wc-post-types/.eslintrc.js:
```diff
@@ -0,0 +1,139 @@
+/*
```
If we use core's eslint rules instead of our own customizations, I think we can avoid adding these .eslintrc files to every sub-project… or maybe we can add it to the top-level of the project? I can look into this more.

iandunn commented:
Blocked per #299, which should be merged first to avoid redoing work.

ryelle committed:
Sessions: Add panel for Speakers to sidebar

iandunn commented:
Removed `blocked` because #299 is merged now.

ryelle committed:
Allow searching all speakers, not just published speakers

ryelle committed:
Move meta/API code into the rest-api file

ryelle committed:
Remove hook call to (already removed) function

ryelle committed:
Add default duration & time

Uses the last-edited session date, defaulting to the WordCamp's date if no sessions exist yet

coreymckrill reviewed (commented):
👍 Worked in local testing

The Session Length number fields I'm seeing aren't sized the same way as in the screenshot; they're way too big and each take up their own line in the panel. Maybe there's a build command somewhere I'm missing? (see screenshot, using Firefox)

![session-length](https://user-images.githubusercontent.com/916023/78617831-7630ea00-782d-11ea-9e38-aa8165529d89.jpg)

The big conundrum I'm facing is if we could merge this given that WordCamp sites can still choose to default to the Classic Editor. Since this PR removes a lot of code necessary for the old metaboxes to work, sessions won't be editable in the Classic view anymore.

Maybe we could add one of those tutorial tool tips pointing at the Session Info panel the first time someone loads an Edit Session screen, so they know where it moved to?

iandunn commented:
> if we could merge this given that WordCamp sites can still choose to default to the Classic Editor.

I think it's fine for us to gradually move support for our _custom_ work to Gutenberg-only. I don't think it's worth it to support both editors for anything custom.

We've already got [code that disables G for some CPTs](https://github.com/WordPress/wordcamp.org/blob/7644a12/public_html/wp-content/mu-plugins/gutenberg-tweaks.php#L20-L37), and [an unfinished PR that expands that in the other direction](https://github.com/WordPress/wordcamp.org/pull/106/commits/622d23a6b868f9dbe8f830db5b6969df8dae9573#diff-addb2c8c6a22887efcd71302eeaaeb94R38-R42).

ryelle commented:
Oh, I forgot to add the build step to my testing instructions 🤦‍♀  This introduces a new project, so you need to run `yarn` & `yarn workspace wc-post-types build`. I'm not sure how you're even seeing any session fields, without the build step.

ryelle commented:
> The big conundrum I'm facing is if we could merge this given that WordCamp sites can still choose to default to the Classic Editor. 

I've got a branch locally to disable the classic editor on select post types, I'll push that up so we can merge it before this.

> Maybe we could add one of those tutorial tool tips pointing at the Session Info panel the first time someone loads an Edit Session screen, so they know where it moved to?

I'll look into this.

ryelle commented:
Looks like the tooltips were deprecated in favor of this style of guide-modal, but I don't think it'll work for us… what do you think?

![Screen Shot 2020-04-07 at 12 41 48 PM](https://user-images.githubusercontent.com/541093/78696244-2ace1880-78cd-11ea-9f99-3145dfe3ef1b.png)

coreymckrill commented:
Hmm, yeah that seems like a bit much. Unless it can be a lot smaller without an image?

ryelle committed:
Fix input widths across browsers

ryelle committed:
Restrict Sessions to only use gutenberg editor

coreymckrill reviewed (approved):
👍 Looks good in Firefox now.

ryelle committed:
Ensure that the session always defaults to having a type

ryelle committed:
Disable (or fix) linting on existing JS files

ryelle committed:
Move files into panel-* folders

ryelle committed:
Update classnames to use full `wordcamp`

Preferred over `wc`, which can be confused with `wp` when scanning

ryelle commented:
Rebased, moved some folders, and fixed an issue with sessions not being correctly tagged as "regular sessions". I don't think this needs a re-review, so will merge when the tests pass.
